Support ST_INTERSECTS between two geometry columns

craigtaverner opened the pull request:
At this point we've only implemented this to do runtime checks of intersection between a geo_point geometry and a polygon geometry. The polygon is converted to a triangle-tree at runtime using Lucene's Polygon2D class. Some important design decisions:

* The first argument must be a correct spatial type: geo_point, cartesian_point, geo_shape, cartesian_shape
* The second argument must be a matching spatial type (same as first)

Starting with the following key use cases:

* Compare literal points to literal polygon:
```
ROW wkt = ["POINT(1 1)", "POINT(-1 -1)", "POINT(-1 1)", "POINT(1 -1)"]
| MV_EXPAND wkt
| EVAL pt = TO_GEOPOINT(wkt)
| WHERE ST_INTERSECTS(pt, TO_GEOSHAPE("POLYGON((0 -1, 1 -1, 1 1, 0 1, 0 -1))"))
```
* Search index of geo_point fields for docs with the point intersecting a literal polygon:
```
FROM airports
| WHERE ST_INTERSECTS(location, TO_GEOSHAPE("POLYGON((42 14, 43 14, 43 15, 42 15, 42 14))"))
```

Tasks to complete before merging this PR:

* [x] Consider whether this is better implemented as a `BinaryComparison` instead of a `BinaryScalarFunction` - No, it is not
* [x] Support cartesian_point
* [x] Support geo_shape
* [x] Support cartesian_shape
* [x] Support push-down to lucene (ie. push the query predicate down to the Lucene query for much improved performance)
* [x] Unit tests for  `SpatialIntersects`
* [x] Verify behaviour against OGC and PostGIS
  * [x] PostGIS does implicit string casting as well, but we should evaluate the scope of this and see if we need to refine our behaviour to better match PostGIS
* [x] Review TODO's
  * [x] Fix TODO related to the use of Exceptions to detect WKB vs WKT (in `SpatialRelations.fromBytesRef()` and `SpatialRelatesTranslator.translate()`)
* [x] Improve test coverage
  * [x] Tests for ST_INTERSECTS before and after ST_CENTROID
  * [x] Tests for ST_INTERSECTS with ENRICH
  * [x] Tests for intersects between two fields of the same index
* [x] Remove support for String parameters to ST_INTERSECTS

Fixes #104874

craigtaverner committed:
Support ST_INTERSECTS between geometry column and other geometry or string

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialIntersects.java:
```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    private static boolean pointIntersectsGeometry(GeoPoint geoPoint, Geometry geometry) {
```
My expectations here is that we have a method that transform a Geometry to a component2D and then we can use the method Component2D#contains. I think we have somewhere else already the code that transform a Geometry to a LatLonGeometry or XYGeometry.

craigtaverner replied:
Now that I've implemented the shape-intersects-shape feature, I have everything necessary to generalize this. However, I think some of these point-specific code paths might still be faster, and worth keeping. We can review this once the code is cleaned up and also once we use your new `LuceneGeometriesUtil`, and done a cleanup because of that.

iverase commented on x-pack/plugin/esql/build.gradle:
```diff
@@ -114,6 +114,7 @@ configurations {
 }
 
 dependencies {
+  implementation project(path: ':x-pack:plugin:spatial')
```
My understanding is that we should not have dependencies between modules

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/EsqlTypeResolutions.java:
```diff
@@ -48,4 +48,8 @@ public static Expression.TypeResolution isExact(Expression e, String operationNa
     }
+
+    public static Expression.TypeResolution isSpatial(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        return isType(e, EsqlDataTypes::isSpatial, operationName, paramOrd, "geo_point or cartesian_point or geo_shape or cartesian_shape");
```
```suggestion
        return isType(e, EsqlDataTypes::isSpatial, operationName, paramOrd, "geo_point", "cartesian_point", "geo_shape", "cartesian_shape");
```

`isType` handles automatically an array of values in a smart way.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelates.java:
```diff
@@ -0,0 +1,225 @@
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.ql.type.DataTypes.isString;
+
+abstract class SpatialRelates extends BinaryScalarFunction implements EvaluatorMapper {
```
I suggest a name having Function in it. `SpatialFunction` for example.

craigtaverner replied:
Done.

craigtaverner committed:
Get geo_shape working in ST_INTERSECTS bypassing SingleValueQuery

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:
```diff
@@ -245,4 +258,91 @@ private static boolean isInRange(DataType numericFieldDataType, DataType valueDa
+         * Skipping this code path entirely is a temporary workaround while separate work is being done to simplify `SingleValueQuery`
+         * to rather rely on a new method on `LeafFieldData`. This is both for the benefit of the spatial queries, as well as an
+         * improvement overall.
+         * TODO: Remove this method and call the parent method once the SingleValueQuery improvements have been made
```
Nik's going to shift the guts of `SingleValueQuery` around a bit to make it more OOP so we can make it compatible with this branch. I'll see if I can do this soon, though it might have to be a followup to this PR.

craigtaverner committed:
Initial work to support cartesian shape queries in ESQL

craigtaverner committed:
Fixed CSV tests for combined ST_INTERSECTS and ST_CENTROID

craigtaverner committed:
Fixed bug in point-in-shape query for CARTESIAN_POINT

craigtaverner committed:
Pushdown to lucene for ST_INTERSECTS on GEO_POINT

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:
```diff
@@ -271,4 +284,91 @@ public static Query doTranslate(ScalarFunction f, TranslatorHandler handler) {
+                    );
+                    return new SpatialRelatesQuery(bc.source(), name, shapeRelation(bc), shape, attribute.dataType());
+                } catch (IllegalArgumentException notWKB) {
+                    // TODO don't use exceptions to detect WKB vs WKT
```
ouch! Don't we know here what is the format? My expectation is that the literal coming from the query is always WKT? I think this needs to be addressed before pushing it

craigtaverner replied:
I will investigate which exact use case caused this, and try remove it.

craigtaverner replied:
There is now a case of this in SpatialRelatesFunction too, and that is part of the implicit string conversion support, so it should go with the next update where we remove implicit conversion of strings.

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java:
```diff
@@ -0,0 +1,271 @@
+            && Objects.equals(dataType, other.dataType);
+    }
+
+    public abstract class ShapeQueryBuilder implements QueryBuilder {
```
Can we add a comment that is public for testing?

craigtaverner replied:
Done

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java:
```diff
@@ -45,10 +45,10 @@ public final class EsqlDataTypes {
-    public static final DataType CARTESIAN_POINT = new DataType("cartesian_point", Double.BYTES * 2, false, false, false);
-    public static final DataType GEO_SHAPE = new DataType("geo_shape", Integer.MAX_VALUE, false, false, false);
-    public static final DataType CARTESIAN_SHAPE = new DataType("cartesian_shape", Integer.MAX_VALUE, false, false, false);
+    public static final DataType GEO_POINT = new DataType("geo_point", Double.BYTES * 2, false, false, true);
```
This field only affects tests, and is currently only used to simulate field-caps behaviour. The real integration tests use field-caps and ignore this field.

craigtaverner committed:
Added unit tests for SpatialIntersects and fixed a few bugs found

craigtaverner committed:
Added comments to public ShapeQueryBuilder class

craigtaverner committed:
Move calls to random() later to avoid security exception

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java:
```diff
@@ -0,0 +1,499 @@
+        GeometryDocValueReader reader = new GeometryDocValueReader();
+        CentroidCalculator centroidCalculator = new CentroidCalculator();
+        if (geometry instanceof Circle circle) {
+            // TODO: How should we deal with circles?
```
You can use the CircleUtils in the geo lib to polygonize it but I think we should throw an error in this case as we do not support indexing circles anyway. shapeIndexer.indexShape(geometry) will throw an error,

craigtaverner replied:
So, since both the centroid calculator and the shape indexer do not support circles, I now throw an exception directly here. Supporting shapes using CircleUtils could be considered as a future enhancement, either by itself, or as part of more advanced Curve support similar to PostGIS.

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java:
```diff
@@ -0,0 +1,499 @@
+        }
+        // Both are spatial, but one could be a literal spatial function
+        if (leftResolution.resolved() && rightResolution.resolved()) {
+            if (left().foldable() && right().foldable() == false || DataTypes.isNull(left().dataType())) {
```
I found very hard to read these conditions, could you add parenthesis to show the order of execution?

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java:
```diff
@@ -0,0 +1,499 @@
+                return SpatialCoordinateTypes.UNSPECIFIED.wkbToGeometry(bytesRef);
+            } catch (IllegalArgumentException e) {
+                // TODO: would be better to not rely on exceptions to tell the difference between WKB and WKT
+                return SpatialCoordinateTypes.UNSPECIFIED.wktToGeometry(bytesRef.utf8ToString());
```
This is not good. Weshould know at every moment if we are dealing with WKT or WKB, otherwise we are loosing control of the situation.

nik9000 replied:
The function should have enough type information to know this. I agree, we should know and call the right one every time.

FWIW at evaluation time we don't annotate the `BytesRefBlock`s with their field type - but we still know the type because that's all set up at planning time.

craigtaverner committed:
Refined type checking support in ST_INTERSECTS

Improved the combinations supported as preparation for removing the uly try/catch way of detecting the difference between WKT and WKB in some code.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialIntersects.java:
```diff
@@ -0,0 +1,274 @@
+        }
+
+        // If both sides are string fields or constants we will treat generically
+        SpatialEvaluatorKey key = SpatialEvaluatorKey.fromSources(DataTypes.KEYWORD, DataTypes.KEYWORD);
```
I'm wondering if we should require the right hand side to be parsed rather than evaluate strings directly.

nik9000 replied:
With, like, `to_geo_point` or whatever.

craigtaverner replied:
I did a review of PostGIS, and they support literal strings. My goal has been to conform to PostGIS functionality whenever possible or reasonable, with the aim of benefitting from PostGIS deep penetration into the market. I plan to do a presentation on the similarities and differences between Spatial-ESQL and PostGIS for us to discuss in a team meeting.

nik9000 replied:
We could infer the conversion and add it rather than push it into the evaluator.

craigtaverner replied:
If we inferred a call to `TO_GEOSHAPE` for example, that would accomplish the first step of the conversion, but we still need to do the DocValues and Component2D conversions. Still this is worth learning more, where would such inference be undertaken?

nik9000 replied:
There's kind of an "easy way" and a "better way". I'd suggest the easy way for now - basically just build the evaluator for the conversion when you build the other evaluator. That's not able to share the conversions, but it does give you a starting point.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java:
```diff
@@ -0,0 +1,580 @@
+            resolveType();
+        }
+        var evaluatorKey = new SpatialEvaluatorKey(crsType, useDocValues, fieldKey(left()), fieldKey(right()));
+        SpatialEvaluatorFactory<?, ?> factory = evaluatorRules().get(evaluatorKey);
```
It'd be nice if the rules were a `static final` map rather than built every time. That's a lot of map entries.

craigtaverner replied:
Yes, that was my plan. I think I forgot to take that last step. Will do so now.

craigtaverner replied:
Actually, the current factory wrappers are not static inner classes, but make use of `source()` and `left()` and `right()` for convenience. Since this map is created at planning time, I do not think this should be too big an issue to create each time a query is planned. If you still think it is better to make this static, I can investigate a way to pass `this` into the class at planning time.

craigtaverner replied:
OK. I've refactored and cleaned it up, so these are now all static.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java:
```diff
@@ -0,0 +1,580 @@
+            return new SpatialEvaluatorKey(crsType, true, left, right);
+        }
+
+        SpatialEvaluatorKey swapSides() {
```
The QL infrastructure already has a fair bit of stuff around constants and rotating constants to the right hand side of an expression. Is that something we can link into here?

craigtaverner replied:
I also think we can do something like what is done in the `AggregateMapper`, where all the various combinations are considered upfront for all aggregations (and a few special cases, like Spatial that I added). I think when I code the next spatial aggregation, and find the type rules are identical to this one, I'll be very tempted to do just that.

craigtaverner replied:
I've improved this in the work to make the mappers static, but it does not yet move this out to whatever QL infrastructure supports this.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialIntersects.java:
```diff
@@ -0,0 +1,274 @@
+    @Evaluator(extraName = "ConstantAndConstant", warnExceptions = { IllegalArgumentException.class, IOException.class })
+    static boolean processConstantAndConstant(@Fixed GeometryDocValueReader leftValue, @Fixed Component2D rightValue) throws IOException {
+        return CARTESIAN.geometryRelatesGeometry(leftValue, rightValue);
+    }
```
Unless geo has some different meaning of the word "constant", I've never seen us have to do this before. The folding infrastructure should kick in if both sides are constant and we should use whatever evaluator is appropriate for the types and it'll be replaced by a constant during the optimization step. But it still runs with a regular evaluator.

craigtaverner replied:
Which part are you commenting on? The use of `@Fixed`, which is something I've seen all over the place when functions can accept foldable types. Or the fact that the left is a `GeometryDocValueReader` and the right is `Component2D`, while the non-constant methods use `BytesRef` encoded WKB? For the second question, the process of figuring out if two geometries are related is to convert one to doc-values (embedded in a `GeometryDocValuesReader` and the other to a `Component2D`, to mimic what Lucene does. For constants we want that conversion done at planning time, hence the use of these types in this method, but for non-constants we will see `BytesRef` WKB formatted geometries coming out the blocks, and need to convert them deeper down the stack.

craigtaverner replied:
OK. I see the `fold()` method, where I was foolishly just called super.fold, but now do the folding directly there, and can remove all these `ConstantAndConstant` methods and generated classes. Definitely an important cleanup.

craigtaverner replied:
OK. That is pushed and working. Definitely much better to use `fold()` directly. Sorry for not spotting this myself!

craigtaverner committed:
Fixed bugs in incorrect use of doc-values in parameter type matching

Also made a few reminfments, including removing one try/catch approach to differentiating between WKT and WKB.

craigtaverner committed:
Removed second place where we used try/catch to differentiate WKT from WKB

This was a workaround for a mistake in the planning, where we incorrectly mapped incoming types to the wrong FieldEvaluators. We fixed that mistake in an earlier commit.

craigtaverner committed:
Fixed flaky tests were GEO was treated as CARTSIAN

We assumed if the incoming types were constants, they had no CRS, even when they did, which was wrong. For shapes crossing the dateline this lead to different (incorrect) behaviour.

craigtaverner committed:
Fixed a flaky test by removing some point==point optimizations

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java:
```diff
@@ -0,0 +1,290 @@
+            case "within" -> ShapeRelation.WITHIN;
+            case "contains" -> ShapeRelation.CONTAINS;
+            default -> throw new QlIllegalArgumentException("Unknown shape relation [{}]", queryRelation);
+        };
```
It is unnecessary to use strings here, you can do it like that? : 

```
 return switch (queryRelation) {
            case INTERSECTS -> ShapeRelation.INTERSECTS;
            case DISJOINT -> ShapeRelation.DISJOINT;
            case WITHIN -> ShapeRelation.WITHIN;
            case CONTAINS -> ShapeRelation.CONTAINS;
        };
```

craigtaverner replied:
Good point. I've been ruined by too many cases where the static type is not a proper `enum` like `DataType` and `EsqlDataType`, where we have to use strings. But since these types are proper `enum` types, I'm fixed this like you suggest.

craigtaverner committed:
Moved spatial intersects to 'spatial' package

When we developed the ST_CENTROID work, this was requested, so let's do it here too.

craigtaverner committed:
Use normal switch on enums

craigtaverner committed:
Cleanup some static utility methods

Now all code paths that can convert a constant string to a geometry use the same code.

craigtaverner committed:
Fixed bugs with non-quantized coordinates, and cleaned up code a little

craigtaverner committed:
Fixed failing test after change to evaluator class names

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialIntersects.java:
```diff
@@ -0,0 +1,288 @@
+    @Evaluator(extraName = "GeoConstantAndConstant", warnExceptions = { IllegalArgumentException.class, IOException.class })
+    static boolean processGeoConstantAndConstant(@Fixed GeometryDocValueReader leftValue, @Fixed Component2D rightValue)
+        throws IOException {
+        return GEO.geometryRelatesGeometry(leftValue, rightValue);
```
I'm pretty sure this shouldn't exist. if all the arguments are `@Fixed` then constant folding should kick in and we'll use one of the other evaluators to make a boolean one time.

craigtaverner replied:
Right, so if both sides are constant, we don't need the generated code, or reading blocks, at all. I know my integration tests hit this code path, so I need to figure out how to make the shortcut happen. I'll investigate.

nik9000 replied:
there's a `foldable` method that *should* work. and then there `fold` method is called - which we generally implement using `EvaluatorMapper.fold` which will land you on a different code path - at or, rather, it'll generate the evaluator and then call it, so it might end up on this code path - but it's not worth optimizing this case then. Because it'll only ever call the evaluator one time.

craigtaverner replied:
I've now pushed a fix that removed all `ConstantAndConstant` methods, classes and generated code, and instead implemented `fold()` correctly. All my tests (unit and integration) are passing, so this does seem like a big improvement.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:
```diff
@@ -0,0 +1,563 @@
+            resolveType();
+        }
+        var evaluatorKey = new SpatialEvaluatorKey(crsType, useDocValues, fieldKey(left()), fieldKey(right()));
+        SpatialEvaluatorFactory<?, ?> factory = evaluatorRules().get(evaluatorKey);
```
`evaluatorRules` should probably create a `static` map if we're going to use it like this.

craigtaverner replied:
Done.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -470,6 +473,15 @@ protected PhysicalPlan rule(AggregateExec aggregate) {
+                if (exec instanceof FilterExec filterExec) {
+                    // Note that ST_CENTROID does not support shapes, but SpatialRelatesFunction does, so when we extend the centroid
+                    // to support shapes, we need to consider loading shape doc-values for both centroid and relates (ST_INTERSECTS)
+                    if (filterExec.condition() instanceof SpatialRelatesFunction spatialRelatesFunction) {
```
Does it matter here if the condition is a boolean composed of multiple other conditions and the spatial function is just one of those conditions?

craigtaverner replied:
This challenge here is that if we optimized ST_CENTROID for reading points as doc-values, then any spatial relates function also needs to expect doc-values, otherwise it fails to interpret the types, since spatial doc-values are completely different to spatial types from source. So this code is making sure that both ST_CENTROID and ST_INTERSECTS know that the data will arrive in a different format.

astefan replied:
What I'm after here (and failed to double check this with an actual query) is the type of the filter you are trying to check. Can `filterExec.condition()` be something like `age > 0 and gender='F' and st_centroid......` ? If so, should you look for a `SpatialRelatesFunction` in the filter and not only check that the filter itself is of type `SpatialRelatesFunction`?

In short

`age > 0 and gender='F' and st_centroid......`
vs
`st_centroid......`

craigtaverner replied:
Right, I see what you mean. I misunderstood your original comment. I will investigate this, and suspect you are right and I'm not dealing with compound boolean expressions correctly.

craigtaverner replied:
I pushed a fix for this. Curiously it only matters for CsvTests, because the lucene-pushdown of ST_INTERSECTS actually erases this scenario.

iverase commented on server/src/main/java/org/elasticsearch/lucene/spatial/CartesianShapeCoordinateEncoder.java:
```diff
@@ -10,7 +10,7 @@
 import org.apache.lucene.geo.XYEncodingUtils;
 
-final class CartesianShapeCoordinateEncoder implements CoordinateEncoder {
+public final class CartesianShapeCoordinateEncoder implements CoordinateEncoder {
```
I would really like to keep them package private, why do you need them?

craigtaverner replied:
These are used in ST_INTERSECTS when creating the geometry doc-values, as well as when creating the Component2DVisitor.

iverase replied:
But they can refer then a `CoordinateEncoder`? whe don't need to made them public?

iverase replied:
You should use CoordinateEncoder.CARTESIAN

iverase replied:
I open https://github.com/elastic/elasticsearch/pull/105502 which hopefully avoid any future issue.

craigtaverner replied:
Understood. I should have used the available constants instead of creating additional classes. I've approved that PR and will switch to that as soon as it's merged.

craigtaverner replied:
Pushed a fix, and use the constants instead now (and removed the `public`)

iverase commented on server/src/main/java/org/elasticsearch/lucene/spatial/GeoShapeCoordinateEncoder.java:
```diff
@@ -11,7 +11,7 @@
 import org.elasticsearch.common.geo.GeoUtils;
 
-final class GeoShapeCoordinateEncoder implements CoordinateEncoder {
+public final class GeoShapeCoordinateEncoder implements CoordinateEncoder {
```
same here: I would really like to keep it package private, why do you need it?

craigtaverner replied:
These are used in ST_INTERSECTS when creating the geometry doc-values, as well as when creating the Component2DVisitor.

iverase replied:
You should use CoordinateEncoder.GEO

craigtaverner replied:
OK. Done.

iverase commented:
I don't agree with what we are trying to implement here. The idea that each argument of the ST_INTERSECTS function can be either a spatial field, a constant or even a text field seems overly complicated and not necessary, at least for a first implementation. In particular I am against in supporting the situation where both arguments do not have a spatial type so we need to magically infer one. This is a behaviour that needs more discussion.

My proposal for this first implementation is to simplify the function arguments. In my opinion the first argument should be always a spatial type, either coming from an Elasticsearch field, or coming from an EVAL via TO_XXX_POINT and TO_XXX_SHAPE functions. The second argument should always be a string literal containing WKT. This eliminates most of the complexity in this PR like trying to guess the contents of each argument and it makes sure we always have spatial type using the fist argument.

Let's please simplify the approach. We can try to cover more cases later on if there is appetite for it.

craigtaverner committed:
Refactored SpatialRelatesFunction into three files, and made evaluatorRules static

This was a general cleanup, making the code more organized, but did also achieve static evaluator rules so we don't re-created these on every query parsing.

craigtaverner committed:
Fixed compile error after rebase

craigtaverner committed:
Removed ConstantAndConstant support, using fold() correctly instead

craigtaverner committed:
better error on circles

craigtaverner committed:
Make sure compound predicates are supported in use-doc-values pushdown

craigtaverner committed:
Testing ENRICH with ST_INTERSECTS

This required adding new data for an ENRICH index, and this data could be tested with a few other related tests, which were also added.

craigtaverner committed:
Added missing mixed-cluster rules for testing only with 8.14

craigtaverner committed:
Fixed some mixed-cluster issues where we failed to mark test for only 8.14

Also added an interesting polygon-polygon intersection case from real data.

craigtaverner commented:
Added an interesting integration test with a administrative region that contains an airport that serves a city where the city location is outside the region. We've got a few tests for this now, but I wanted to add the image for a polygon-polygon intersection we test here too.
<img width="631" alt="Screenshot 2024-02-23 at 19 31 47" src="https://github.com/elastic/elasticsearch/assets/114451/e49cb954-ddbb-44d1-8803-7299ee99c814">

craigtaverner committed:
Fix flaky test where cartesian polygons were generated from geo

craigtaverner committed:
Remove support for string literals in ST_INTERSECTS

craigtaverner committed:
Fix failing tests after removing string support

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesUtils.java:
```diff
@@ -0,0 +1,112 @@
+        if (value instanceof BytesRef bytesRef) {
+            if (EsqlDataTypes.isSpatial(expr.dataType())) {
+                return SpatialCoordinateTypes.UNSPECIFIED.wkbToGeometry(bytesRef);
+            } else {
```
I think the string parsing is not needed anymore (and those branches are not being exercise by test as expected.). Could be remove those branches from here to simplify it?

craigtaverner replied:
Yes, we don't need those anymore. I'm removing them now.

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:
```diff
@@ -0,0 +1,280 @@
+            }
+        }
+        // If any parameter is not spatial, we are unresolved
+        if ((leftResolution.unresolved() || isNull(left().dataType())) && (rightResolution.unresolved() || isNull(right().dataType()))) {
```
This is not supported anymore? let's remove it?

craigtaverner replied:
We need this for the type error generation. When the function is called with the wrong types, it generates an error message explaining what types it supports and what type the user incorrectly used. These lines are mostly about correct error generation. It is made more complex by the need to support nulls, because if we call it with a null and a spatial type, or two nulls, we don't generate any error, but simply return a null result. But if one type is null, we should generate an error based on the other type. The SpatialIntersectsTests do extensive testing of all possible type combinations and verify that all the correct errors are generated.

nik9000 replied:
I wonder if we're better off handling the null cases up front. Honestly this comment helps explain what's happening better than the comments in the code. A little table in the javadoc could help. Something like that.

Also - are we sure that if one of the two is `null` we shouldn't just return `null`? Most other ESQL functions will return `null` if either parameter is `null`. We even have a test utility for it - `anyNullIsNull`. IBM seems to do it that way: https://www.ibm.com/docs/en/ias?topic=functions-st-intersects

That's a lot easier to write too.

costin replied:
👍 to Nik's reply to keep things simple - if either side is null, null should be returned.
I understand how the code got here however, extracting the null stuff out will make the code larger but more readable.
Note that there's a subtle difference between null (as in value) vs null as datatype - the latter is rarely used in ESQL and we might forgo it all together.
That is fields always have a type expect except when the user specifies `null` which we could handle through implicit type conversion so that the functions don't have to handle it.

iverase commented:
This does not seems to work:
```
PUT /geo?pretty
{
  "mappings": {
    "properties": {
      "location1": {
        "type": "geo_point"
      },
      "location2": {
        "type": "geo_point"
      }
    }
  }
}

POST geo/_doc
{
  "location1" : "0,0",
  "location2" : "0,0"
}

POST /_query
{
  "query": """
   from geo |
   where ST_INTERSECTS(location1, location2)
   | STATS ST_CENTROID(location2)
  """
}
```

fails with error:

```
 "type": "class_cast_exception",
        "reason": "class org.elasticsearch.compute.data.BytesRefVectorBlock cannot be cast to class org.elasticsearch.compute.data.LongBlock (org.elasticsearch.compute.data.BytesRefVectorBlock and org.elasticsearch.compute.data.LongBlock are in unnamed module of loader java.net.FactoryURLClassLoader @595fed99)"
```

craigtaverner committed:
Removed unused code from previous string literal support (WKT parsing)

craigtaverner commented:
> This does not seems to work:
> ```
>  "type": "class_cast_exception",
>         "reason": "class org.elasticsearch.compute.data.BytesRefVectorBlock cannot be cast to class org.elasticsearch.compute.data.LongBlock (org.elasticsearch.compute.data.BytesRefVectorBlock and org.elasticsearch.compute.data.LongBlock are in unnamed module of loader java.net.FactoryURLClassLoader @595fed99)"
> ```

So this is because the ST_CENTROID asks for the right field to be extracted using doc-values. If I use STATS to get centroids for both fields, they are both extracted as doc-values. The problem is that the ST_INTERSECTS is only setup to support one doc values field, the left. I'm making a small addition to support both, and will create unit and regressions tests for this case.

craigtaverner committed:
Support case where both fields are points and doc-values

If we have an ST_INTERSECTS and an ST_CENTROID, the centroid asks to load the points as doc-values, and the ST_INTERSECTS needs to therefor support two doc-values points.

craigtaverner committed:
Disallow more than one field from doc-values for ST_INTERSECTS

craigtaverner committed:
Remove unused evaluator classes

craigtaverner committed:
Add tests for multiple doc-values if not in same intersects

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/EsqlTypeResolutions.java:
```diff
@@ -46,6 +46,10 @@ public static Expression.TypeResolution isExact(Expression e, String operationNa
 
     public static Expression.TypeResolution isSpatialPoint(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
-        return isType(e, EsqlDataTypes::isSpatialPoint, operationName, paramOrd, "geo_point or cartesian_point");
+        return isType(e, EsqlDataTypes::isSpatialPoint, operationName, paramOrd, "geo_point", "cartesian_point");
+    }
+
+    public static Expression.TypeResolution isSpatial(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        return isType(e, EsqlDataTypes::isSpatial, operationName, paramOrd, "geo_point", "cartesian_point", "geo_shape", "cartesian_shape");
```
nit pick - how about extracting these strings into constants to avoid typos ("geo_poitn")

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -474,6 +478,20 @@ protected PhysicalPlan rule(AggregateExec aggregate) {
+                if (exec instanceof FilterExec filterExec) {
+                    // Note that ST_CENTROID does not support shapes, but SpatialRelatesFunction does, so when we extend the centroid
+                    // to support shapes, we need to consider loading shape doc-values for both centroid and relates (ST_INTERSECTS)
+                    var condition = filterExec.condition()
+                        .transformDown(
+                            SpatialRelatesFunction.class,
+                            spatialRelatesFunction -> (spatialRelatesFunction.hasFieldAttribute(foundAttributes))
+                                ? spatialRelatesFunction.withDocValues(foundAttributes)
+                                : spatialRelatesFunction
+                        );
+                    if (filterExec.condition().equals(condition) == false) {
+                        exec = new FilterExec(filterExec.source(), filterExec.child(), condition);
+                    }
+                }
```
it makes sense to do this check if you can easily track if something was modified - if you can't and rely on objection comparison, just rely on transform methods which already perform this:
`return filterExec.transformExpressionOnly(SpatialRelatesFunction.class, ...)`

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlExpressionTranslators.java:
```diff
@@ -240,4 +247,63 @@ public static Query doTranslate(ScalarFunction f, TranslatorHandler handler) {
+         * Skipping this code path entirely is a temporary workaround while separate work is being done to simplify `SingleValueQuery`
+         * to rather rely on a new method on `LeafFieldData`. This is both for the benefit of the spatial queries, as well as an
+         * improvement overall.
+         * TODO: Remove this method and call the parent method once the SingleValueQuery improvements have been made
```
Since this is outside this PR, please raise a GH issue and point to it.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java:
```diff
@@ -0,0 +1,287 @@
+
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.CARTESIAN_POINT;
+
+public class SpatialRelatesQuery extends Query {
```
Reminder to self that moving to Lucene queries should simplify things.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:
```diff
@@ -108,7 +108,19 @@ public BytesRef wktToWkb(String wkt) {
+        try {
+            return WellKnownText.fromWKT(GeometryValidator.NOOP, false, wkt);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Failed to parse WKT: " + e.getMessage(), e);
```
Use InvalidArgumentException instead (through-out the code as well).

craigtaverner replied:
Interesting. I took a look around, and found many more usages of IllegalArgumentException in the csv-spec tests and related functions. Looks like InvalidArgumentException is used only in a few places. Are we planning to change this all over to InvalidArgumentException? Perhaps that should be done as a separate PR for all of these cases, since it is much wider scope that this PR?

costin reviewed (approved):
It took me a while to review this PR - sorry.
Can't comment on the Geo bits but everything else looks in order to me hence why I'm +1.
Left some minor methods but nothing major.

craigtaverner committed:
Fix errors after rebase on main

craigtaverner committed:
Fixed bug in missing support for spatial function expressions in EVAL

When a spatial aggregate expects doc-values, this was not being communicated to spatial functions in EVAL, only in WHERE.

craigtaverner committed:
Reduce flaky tests when reading directly from enrich source indices

The test framework does not expect enrich source indices to be used directly in queries, leading to duplicated results on multi-node clusters, so we edit the queries to be less sensitive to this case.

craigtaverner committed:
Fixed failing test

craigtaverner committed:
Code style

craigtaverner committed:
Fixed test file name and added function name annotation

craigtaverner committed:
Added documentation for st_intersects

craigtaverner committed:
Fixed failing show functions test

craigtaverner committed:
Code review changes, notably simplifying the type resolution

craigtaverner committed:
Fixed broken docs link

bouyeijiang commented:
Does it support the intersection of spatial columns between two tables?

craigtaverner commented:
> Does it support the intersection of spatial columns between two tables?

Well, that would be a spatial join, and JOIN itself is not yet supported by ES|QL. However, there is a special subset of JOIN called [ENRICH](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-commands.html#esql-enrich) which can be used for some cases. Consider the following two examples:

* [ENRICH on keyword followed by ST_INTERSECTS filtering](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/testFixtures/src/main/resources/enrich-IT_tests_only.csv-spec#L309-L314)
* [ENRICH on spatial intersects](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/testFixtures/src/main/resources/enrich-IT_tests_only.csv-spec#L280-L284)

You can click on the links to see the sample code in the integration tests, but for clarity I've copied it here. The first one is:

```
FROM airports
| ENRICH city_names ON city WITH airport, region, city_boundary
| MV_EXPAND city_boundary
| EVAL airport_in_city = ST_INTERSECTS(location, city_boundary)
| STATS count=COUNT(*) BY airport_in_city
| SORT count ASC
```

This does a kind-of JOIN between the `airports` table and the `city_names` table by matching the airports city name fields between the two tables. This requires that the city_names ENRICH index be [pre-configured for ENRICH on the city name](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/testFixtures/src/main/resources/enrich-policy-city_names.json#L2-L6) field.

The second example does not explicitly have an `ST_INTERSECTS` call, but is closer to the use case you are after, because `ENRICH` internally supports `geo_match`:

```
FROM airports
| WHERE abbrev == "CPH"
| ENRICH city_boundaries ON city_location WITH airport, region, city_boundary
| EVAL boundary_wkt_length = LENGTH(TO_STRING(city_boundary))
| KEEP abbrev, city, city_location, country, location, name, airport, region, boundary_wkt_length
```

The ENRICH policy was [preconfigured to be a `geo_match` policy](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/testFixtures/src/main/resources/enrich-policy-city_boundaries.json#L2-L6) with `intersects` by default:

```
{
  "geo_match": {
    "indices": "airport_city_boundaries",
    "match_field": "city_boundary",
    "enrich_fields": ["city", "airport", "region", "city_boundary"]
  }
}
```

So while the short answer is "No we cannot perform spatial JOIN", the long answer is "Yes, we can under some conditions with ENRICH". The "long" part of the answer is the preparation of [ENRICH policies](https://www.elastic.co/guide/en/elasticsearch/reference/current/put-enrich-policy-api.html) appropriate to your use case.

bouyeijiang commented:
谢谢您的来信！我会及时查阅的
 

bouyeijiang commented:
Okay,this is also a method.Maybe I should study the usage of ENRICH policies
